### PR TITLE
Add TFLite benchmarks for formerly TF models

### PR DESCRIPTION
**Pull request by GMNGeoffrey**
These benchmarks align very well with the existing TF source benchmarks
but don't involve going through our less-well-supported TF source
integration or storing unstable MLIR artifacts.

I did a comparison of benchmarks before and after on my Pixel 4 dev
phone, manually rewriting the model source in the former so that a
direct comparison was possible with existing tooling. No benchmarks
have significant changes:
https://gist.github.com/GMNGeoffrey/bce029bf4697f9b3deda3bb217b0c6b3

### CODE REVIEW

**Comment by GMNGeoffrey** [comment]
Seems the benchmarking isn't very fault-tolerant:

```
[2021-11-12T18:13:19Z] I ../iree/tools/utils/vm_util.cc:185] Creating driver and device for 'vulkan'...
--
  | [2021-11-12T18:13:24Z] ../iree/hal/vulkan/serializing_command_queue.cc:271: INTERNAL; VK_ERROR_DEVICE_LOST; vkQueueSubmit; while invoking native function hal.ex.submit_and_wait; while calling import;
```

That's a transient error, I'm pretty sure. Probably we should add retry logic and also make output incremental, so we still get results even if one benchmark fails. I'm going to retry the step on buildkite for now

**Comment by GMNGeoffrey** [comment]
> Seems the benchmarking isn't very fault-tolerant:
> 
> ```
> [2021-11-12T18:13:19Z] I ../iree/tools/utils/vm_util.cc:185] Creating driver and device for 'vulkan'...
> --
>   | [2021-11-12T18:13:24Z] ../iree/hal/vulkan/serializing_command_queue.cc:271: INTERNAL; VK_ERROR_DEVICE_LOST; vkQueueSubmit; while invoking native function hal.ex.submit_and_wait; while calling import;
> ```
> 
> That's a transient error, I'm pretty sure. Probably we should add retry logic and also make output incremental, so we still get results even if one benchmark fails. I'm going to retry the step on buildkite for now

The error was not transient. It has to do with the kernel-execution method of making giant dispatches that was going over the timeout for a single dispatch. Fixed and added a comment about this pitfall.

**Comment by antiagainst** [comment]
This is marked as draft and still have failures but I see you requested review. Is it ready for review now or not? :)

**Comment by GMNGeoffrey** [comment]
> This is marked as draft and still have failures but I see you requested review. Is it ready for review now or not? :)

Argh. My intent was that it was ready, but looks like we've still got a failing benchmark :-( I thought there was something funky with my local device, since I'm pretty sure these had worked previously

```
[2021-11-17T02:56:29Z] cmd: adb shell cd /data/local/tmp/iree-benchmarks/TFLite/MobileBertSquad-fp32/iree-vulkan__GPU-Adreno__kernel-execution,experimental-flags && taskset 80 /data/local/tmp/iree-benchmarks/normal-tools/iree-benchmark-module --flagfile=flagfile --benchmark_repetitions=10 --benchmark_format=json
--
  | [2021-11-17T02:56:29Z] I ../iree/tools/utils/vm_util.cc:185] Creating driver and device for 'vulkan'...
  | [2021-11-17T02:56:44Z] ../iree/hal/vulkan/serializing_command_queue.cc:271: INTERNAL; VK_ERROR_DEVICE_LOST; vkQueueSubmit; while invoking native function hal.ex.submit_and_wait; while calling import;
  | [2021-11-17T02:56:44Z] [ 1]   native hal.ex.submit_and_wait:0 -
  | [2021-11-17T02:56:44Z] [ 0] bytecode module.main:320498 /usr/src/github/iree/build-host/benchmark_suites/TFLite/mobilebertsquad.tflite.mlir:2:3
  | [2021-11-17T02:56:44Z] Aborted
  | [2021-11-17T02:56:45Z] Traceback (most recent call last):
  | [2021-11-17T02:56:45Z]   File "build_tools/benchmarks/run_benchmarks_on_android.py", line 553, in <module>
  | [2021-11-17T02:56:45Z]     main(parse_arguments())
  | [2021-11-17T02:56:45Z]   File "build_tools/benchmarks/run_benchmarks_on_android.py", line 519, in main
  | [2021-11-17T02:56:45Z]     results, captures = filter_and_run_benchmarks(
  | [2021-11-17T02:56:45Z]   File "build_tools/benchmarks/run_benchmarks_on_android.py", line 415, in filter_and_run_benchmarks
  | [2021-11-17T02:56:45Z]     run_results = run_benchmarks_for_category(
  | [2021-11-17T02:56:45Z]   File "build_tools/benchmarks/run_benchmarks_on_android.py", line 329, in run_benchmarks_for_category
  | [2021-11-17T02:56:45Z]     resultjson = adb_execute_in_dir(cmd, android_relative_dir, verbose=verbose)
  | [2021-11-17T02:56:45Z]   File "build_tools/benchmarks/run_benchmarks_on_android.py", line 143, in adb_execute_in_dir
  | [2021-11-17T02:56:45Z]     return execute_cmd_and_get_output(cmd, verbose=verbose)
  | [2021-11-17T02:56:45Z]   File "/var/lib/buildkite-agent/builds/IREE-LAB-RPI4-PIXEL4-ubuntu-1/iree/iree-benchmark/build_tools/benchmarks/common/benchmark_definition.py", line 44, in execute_cmd_and_get_output
  | [2021-11-17T02:56:45Z]     return subprocess.run(args,
  | [2021-11-17T02:56:45Z]   File "/usr/lib/python3.8/subprocess.py", line 516, in run
  | [2021-11-17T02:56:45Z]     raise CalledProcessError(retcode, process.args,
  | [2021-11-17T02:56:45Z] subprocess.CalledProcessError: Command '['adb', 'shell', 'cd', '/data/local/tmp/iree-benchmarks/TFLite/MobileBertSquad-fp32/iree-vulkan__GPU-Adreno__kernel-execution,experimental-flags', '&&', 'taskset', '80', '/data/local/tmp/iree-benchmarks/normal-tools/iree-benchmark-module', '--flagfile=flagfile', '--benchmark_repetitions=10', '--benchmark_format=json']' returned non-zero exit status 134.
```

That's a GPU timeout again, right? For some reason I thought I'd already fixed it. I'll get everything working and then mark ready for review. Sorry for the noise

**Comment by GMNGeoffrey** [comment]
Ok had to remove the mobile bert kernel benchmark and now we're good to go

**Review by antiagainst — CHANGES_REQUESTED** [review]
Nice, thanks! Generally LGTM; just need to remove VMVX 3-thread for now.

File: benchmarks/TFLite/CMakeLists.txt [hunk 1]
```diff
@@ -217,6 +300,27 @@ iree_benchmark_suite(
...
+    "${MOBILEBERT_FP16_MODULE}"
+
+  BENCHMARK_MODES
+    "full-inference,default-flags"
+  TARGET_BACKEND
+    "vulkan-spirv"
+  TARGET_ARCHITECTURE
+    "GPU-Mali-Valhall"
+  TRANSLATION_FLAGS
+    "--iree-input-type=tosa"
```
[the remark this answers is not in the document]
Comment (antiagainst) [reply benchmarks/TFLite/CMakeLists.txt:315]: (This can be deferred to a later commit. We can also sort the models by name there.)

**Comment by antiagainst** [comment]
Also, with this we need to update docs [here](https://github.com/google/iree/blob/main/benchmarks/README.md) and [here](https://github.com/google/iree/blob/main/benchmarks/dashboard.md). But again, can happen later. :)

**Comment by benvanik** [comment]
out of curiosity, is there an easy way to get the MLIR artifacts from these without needing to build tf? I'm really excited to not have the unstable MLIR artifacts as a source, but not excited to need to build iree-import-tflite and construct command lines to repro benchmarks locally (or do random colab hacking or python release shenanigans). Is there a script that could fetch them from somewhere or at least automate the build + convert step for local use?

**Comment by benvanik** [comment]
for reference, yesterday I clicked the tgz link in the CMakeLists.txt and had a .mlir I could immediately pass to the compiler to repro - not expecting something that easy but something closer to that then "hope you have something in your bash history from the last time you did this a month ago and spend an hour building tf" :P

**Comment by antiagainst** [comment]
The `iree-benchmark-suites` target is meant for both the CI and local use. It should download all necessary stuff from the cloud and generate all artifacts. I think we can relax the following a bit:

https://github.com/google/iree/blob/66b8d7aaae8dc7fdb6e648a3116e7d237aeda872/build_tools/cmake/iree_benchmark_suite.cmake#L136-L139

To let it also look at where `iree-improt-tfilte` is typically installed via pip so that you can install it via pip?

**Comment by benvanik** [comment]
that'd be useful! it still requires pip stuff and an up-to-date release, but I feel like the tflite layer changes relatively infrequently so it'd probably be fine if that's all it needs from the package and can use the rest (iree-translate/etc) from the local build

**Comment by GMNGeoffrey** [comment]
@benvanik you can use the IREE tools from snapshot releases to generate them from the TFLite flatbuffers

> The `iree-benchmark-suites` target is meant for both the CI and local use. It should download all necessary stuff from the cloud and generate all artifacts. I think we can relax the following a bit:
> 
> https://github.com/google/iree/blob/66b8d7aaae8dc7fdb6e648a3116e7d237aeda872/build_tools/cmake/iree_benchmark_suite.cmake#L136-L139
> 
> To let it also look at where `iree-improt-tfilte` is typically installed via pip so that you can install it via pip?

You'd still want to set `IREE_BUILD_TFLITE_COMPILER` here, just point it to the place where the pip install is (this is true of all the stuff with the TF binaries.

The benchmark runs also upload artifacts. They currently exclude .mlir files (https://github.com/google/iree/blob/c3505092c2ec709c86e9cfe73fc68b06d730c70e/build_tools/buildkite/cmake/android/arm64-v8a/benchmark2.yml#L13), but we could have them included (in fact, we should probably be excluding .tflite files there!). Then the most up to date could be downloaded from the latest benchmark run

**Comment by benvanik** [comment]
> You'd still want to set IREE_BUILD_TFLITE_COMPILER here, just point it to the place where the pip install is

not familiar with this stuff, but isn't that a boolean flag? how do I point it at things? (I'd assume if I said IREE_BUILD_TFLITE_COMPILER=ON it would build the tflite compiler, which is what I want to avoid?)

I think it's mostly about the process - needing to do a bunch of commands that may fail to then type out a bunch of commands that may fail and run them in turn all sucks, but it sounds like if we can make the target work it'll be pretty nice. It'll also be useful for people who want to run the benchmarks locally on their own hardware!

**Comment by antiagainst** [comment]
> The benchmark runs also upload artifacts. They currently exclude .mlir files (
> 
> https://github.com/google/iree/blob/c3505092c2ec709c86e9cfe73fc68b06d730c70e/build_tools/buildkite/cmake/android/arm64-v8a/benchmark2.yml#L13
> 
> ), but we could have them included (in fact, we should probably be excluding .tflite files there!). Then the most up to date could be downloaded from the latest benchmark run

Good catch! That should be fixed. Yeah. That works too. But I always have the tarball size concern, esp. this is uploading stuff each commit.. Do you know how much they total? I'd assume ~100MB?

**Comment by antiagainst** [comment]
> I think it's mostly about the process - needing to do a bunch of commands that may fail to then type out a bunch of commands that may fail and run them in turn all sucks, but it sounds like if we can make the target work it'll be pretty nice. It'll also be useful for people who want to run the benchmarks locally on their own hardware!

Yup. That's the original intent behind  the `iree-benchmark-suites` target. I basically use it locally for development.

File: benchmarks/TFLite/CMakeLists.txt [hunk 2]
```diff
@@ -346,12 +465,38 @@ iree_benchmark_suite(
     "--task_topology_group_count=4"
 )
 
+
+# CPU, VMVX, 3-thread, little-core, full-inference
```
[the remark this answers is not in the document]
Comment (GMNGeoffrey) [reply benchmarks/TFLite/CMakeLists.txt]: Chatted offline. This is our only VMVX benchmark and we agreed it's good to have one. Switched to 4 threads (was only using 3 for consistency with the previous TF-based benchmark, but I don't think we care too much about the history here). Added a comment about the usefulness of this benchmark.

File: benchmarks/TFLite/CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (GMNGeoffrey) [reply benchmarks/TFLite/CMakeLists.txt]: Good idea, done

**Comment by GMNGeoffrey** [comment]
> > You'd still want to set IREE_BUILD_TFLITE_COMPILER here, just point it to the place where the pip install is
> 
> not familiar with this stuff, but isn't that a boolean flag? how do I point it at things? (I'd assume if I said IREE_BUILD_TFLITE_COMPILER=ON it would build the tflite compiler, which is what I want to avoid?)

Yeah that one's a boolean flag. There's also `IREE_TF_TOOLS_ROOT` which tells CMake where to look for the already-built binaries (https://github.com/google/iree/blob/c3505092c2ec709c86e9cfe73fc68b06d730c70e/integrations/tensorflow/iree_tf_compiler/CMakeLists.txt#L7). Yeah that option name is a bit confusing... Probably should be "enable" instead. I think the name may be a holdover from when Stella put Bazel inside of CMake, so this actually controlled doing a Bazel invocation :fearful: Now it hooks in the binary (https://github.com/google/iree/blob/c3505092c2ec709c86e9cfe73fc68b06d730c70e/integrations/tensorflow/iree_tf_compiler/CMakeLists.txt#L30) and switches on TF stuff, including building the Python bindings (https://github.com/google/iree/blob/c3505092c2ec709c86e9cfe73fc68b06d730c70e/CMakeLists.txt#L61-L74). Probably we should decouple the python bindings from that. Not really sure why those need to be coupled.

**Comment by GMNGeoffrey** [comment]
> Good catch! That should be fixed. Yeah. That works too. But I always have the tarball size concern, esp. this is uploading stuff each commit.. Do you know how much they total? I'd assume ~100MB?

Looking at my local artifacts, they're 286MB uncompressed, 155MB in a tar.gz. At some point we could get fancy and upload in such a way that we only reupload if they've changed. Right now I think the option is on or off though :-)

**Review by antiagainst — APPROVED** [review]
Thanks for migrating these over!!!

**Comment by ScottTodd** [comment]
Could you please update the documentation at https://github.com/google/iree/blob/main/benchmarks/README.md and https://github.com/google/iree/blob/main/benchmarks/dashboard.md following this change? The `TensorFlow/` folder no longer exists, and `.tflite` files are used as input instead of `.mlir` files.

(I'd also like to see the discussion from last night about big core vs little cores and other benchmarking methodology written down, perhaps in https://github.com/google/iree/blob/main/benchmarks/dashboard.md#benchmarking-philosophy, but that's a separate topic)

**Pull request merged**